### PR TITLE
Adjust disambiguation rules for type pattern in a switch expression.

### LINKED ISSUE #47614: Type pattern in switch expression involving generic type with non-keyword type parameters causes syntax error
**Version Used**: Compiler version: '3.8.0-4.20460.4 (64f2392e)'. Language version: preview.

**Steps to Reproduce**:

Compile this code: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFBWEDMABCfQCpyxUDeV9PjDhpAAz0AsgAoA9sABWcAMYx6EgJTde6gLwA+JfSgB3TDDkALNep5dKFm4yH1t9AEQBlGAkwA7AOZOk52x4BYgcdJ2C/AMDggB4vGB1HJwBBOSMJTzjPBMjrQN5YqVkFRLDCmXkc/zz8u2IYgEls0gA2UudYpphWrVzaoNJ61lh28MGY4aqo2wB9UOcAVU8Aa08JfU8naZ4AXwBuKh2qIA))

```csharp
using System;

public class Test
{
    public string M(object o)
        => o switch
        {
            string => "String",
            Action => "Action",
            Action<int> => "Aciton<int>",
            Action<object> => "Action<object>",
            Action<Int16> => "Action<Int16>", // <-- Syntax Errors
            Action<Test> => "Action<Test>", // <-- Syntax Errors
            _ => "Unknown"
        };
}
```

**Expected Behavior**: Compiles without errors

**Actual Behavior**: The lines marked have errors associated with them:

```
Line 12: error CS1001: Identifier expected
Line 12: error CS1003: Syntax error, '=>' expected
Line 12: error CS1525: Invalid expression term ','
Line 13: error CS1001: Identifier expected
Line 13: error CS1003: Syntax error, '=>' expected
Line 13: error CS1525: Invalid expression term ','
```

-------------

This does not happen with the equivalent switch statement: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFBWEDMABCfQCpyxUDeV9PjDhpAAz0AsgAoA9sABWcAMYx6EgJTdeXSry30oAd0ww5AC3qTVm7Tw2XLc6HEZCQamzcIB2egCIAyjASYAHYA5l4A3C6u9HZQDgLEzhZR2h7e8eGRrjFxpMQAPEEwAHyJyZapXgCCcgYSgQWBxRlJUdmO+VKyCiWZZRXVtfWd8k0RLVn27XkAko2kAGw9432eVTUwdTNzi81l0ZPxeaywS3ta/eubx6O9lgAmcNgQAK4ANqi3URUAqoEA1oEJLpArsbABfSIQyhQoA))

```csharp
using System;

public class Test
{
    public string M(object o)
    {
        switch (o)
        {
            case string:
                return "String";
            case Action:
                return "Action";
            case Action<int>:
                return "Aciton<int>";
            case Action<object>:
                return "Aciton<object>";
            case Action<Int16>:
                return "Aciton<Int16>";
            case Action<Test>:
                return "Aciton<Test>";
            default:
                return "Unknown";
        }
    }
}
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #47614
See also https://github.com/dotnet/csharplang/commit/2661a4b3950d2ec5c557b025af13e52ba200fd05

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs [hunk 1]
```diff
@@ -11211,5 +11211,86 @@ void verifyTree()
...
+                            }
+                        }
+                        N(SyntaxKind.CommaToken);
+                        N(SyntaxKind.CloseBraceToken);
+                    }
+                }
+                N(SyntaxKind.SemicolonToken);
+            }
+            EOF();
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs:11294]: Could you also add a CompileAndVerify test?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1) modulo a execution test with generic type

File: src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs:11294]: We already have tests for this kind of scenario but with types that are not precisely the problematic syntax. This is purely a parsing problem.

File: src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs [hunk 2]
```diff
@@ -11211,5 +11211,86 @@ void verifyTree()
                 EOF();
             }
         }
+
+        [Fact, WorkItem(47614, "https://github.com/dotnet/roslyn/issues/47614")]
+        public void GenericTypeAsTypePatternInSwitchExpression()
+        {
+            UsingStatement(@"_ = e switch { List<Foo> => 1, List<Bar> => 2, };",
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs:11218]: >Foo [](start = 49, length = 3)

Consider using different names, perhaps "Int32", "Object"

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests.cs:11218]: Yeah. Oops.

**Pull request merged**